### PR TITLE
Add DataSource and TypedLocalObjectReference

### PR DESCRIPTION
**Pull request by xing-yang**
**What this PR does / why we need it**:
This PR adds TypedLocalObjectReference in the core API and adds DataSource in PersistentVolumeClaimSpec.

It also enables feature gate for VolumeSnapshotDataSource.

This is part of the CSI snapshot design proposal to support restoring a volume from a snapshot: 
kubernetes/community#2495

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
kubernetes/features#177

**Special notes for your reviewer**:

**Release note**:

```release-note
Added support to restore a volume from a volume snapshot data source. 
```

### CODE REVIEW

**Comment by xing-yang** [comment]
@jingxu97

**Comment by xing-yang** [comment]
@saad-ali

**Comment by saad-ali** [comment]
/lgtm
Looks good from sig-storage side.
Needs API Approval. CC @kubernetes/api-approvers 
/assign @thockin

File: pkg/apis/core/types.go
Comment (thockin) [inline pkg/apis/core/types.go]: @lavalamp @kubernetes/sig-api-machinery-api-reviews  I need a consult.

Is "kind" sufficient here?  In general a kind holds something like "Endpoints" while apiVersion holds "extensions/v1beta1".  In a context like this, I don't care which API version, but I do care which group.

**Review by thockin — COMMENTED** [review]
This PR is missing the readiness gate portion.  Do you really want to merge this without that?

The risk is that you don't get the readiness done in time, and the release cuts with a half-cooked feature.  We're trying to do less in-progress development in master to avoid this problem.

Should this wait for the readiness feature to be mergeable?  Either as part of this PR or as a pre-req for this.

File: pkg/apis/core/types.go
Comment (liggitt) [reply pkg/apis/core/types.go]: kind, API group, and name are the three bits of info needed to unambiguously reference a particular object (assuming you don't care what version of the object you deal with).

for example:
https://github.com/kubernetes/kubernetes/blob/7749eeaf3f1eddd0796a4d61e545a90e7d9d59b0/staging/src/k8s.io/api/rbac/v1/types.go#L90-L98

**Comment by xing-yang** [comment]
@thockin We have another PR on the readiness gate depending on this PR: https://github.com/kubernetes/kubernetes/pull/67625.  Do you want us to combine them into one PR?  Thanks.

File: pkg/apis/core/types.go
Comment (thockin) [reply pkg/apis/core/types.go]: Thanks, Jordan, that's what I assumed.  I saw that ObjectReference has a field "APIVersion" which seemed wrong...  I didn't know this struct, but it makes more sense.

**Comment by thockin** [comment]
@xing-yang The question is whether we collectively are willing to launch 1.12 with this DataSource but without readiness gates.  My guess is no?

**Comment by xing-yang** [comment]
@thockin Based on the concerns you raised, we believe that it is important to include the readiness gate part in this PR.  We'll make that change soon.  Thanks!

File: pkg/features/kube_features.go
Comment (jingxu97) [inline pkg/features/kube_features.go]: We need to add validation for data source. The kind, API group etc.

Comment (xing-yang) [reply pkg/features/kube_features.go]: Sure

File: pkg/apis/core/types.go
Comment (xing-yang) [reply pkg/apis/core/types.go]: Added APIGroup.

Comment (jingxu97) [reply pkg/apis/core/types.go]: Then we will add the APIGroup in the struct. I assumed it is required and user has to specify the APIGroup here for the data source.

File: pkg/features/kube_features.go
Comment (thockin) [reply pkg/features/kube_features.go]: Somewhere you need to test the field gate and nil the field if it is not enabled.  See other alpha fields for examples.

**Comment by thockin** [comment]
Ping me when this is ready

**Comment by xing-yang** [comment]
@thockin we are adding a validation.  Will ping you when it’s ready.  Thanks.

File: pkg/features/kube_features.go
Comment (xing-yang) [reply pkg/features/kube_features.go]: Done

File: pkg/apis/core/types.go
Comment (thockin) [inline pkg/apis/core/types.go]: need to document something like "This depends on the provisioner for this volume being able to use the specified source.  If the provisioner does not support it, ...:

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: Don't repeate types and fields in validation errors - those are already provided.  Please follow style of others in this file.

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: "only VolumeSnapshot is supported"

consider field.NotSupported() here

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: no need to set to nil if returning an error?

File: pkg/apis/core/types.go
Comment (xing-yang) [reply pkg/apis/core/types.go]: Done

File: pkg/apis/core/validation/validation.go
Comment (xing-yang) [reply pkg/apis/core/validation/validation.go]: Done

File: pkg/apis/core/validation/validation.go
Comment (xing-yang) [reply pkg/apis/core/validation/validation.go]: Done

File: pkg/apis/core/validation/validation.go
Comment (xing-yang) [reply pkg/apis/core/validation/validation.go]: Remove

**Comment by xing-yang** [comment]
@thockin Addressed you comments.  Please review again.  Thanks!

**Comment by jingxu97** [comment]
@xing-yang, PR needs rebase?

**Comment by xing-yang** [comment]
Rebased.

File: pkg/apis/core/types.go
Comment (jingxu97) [inline pkg/apis/core/types.go]: Need more details here, see Tim's comment in the proposal https://github.com/kubernetes/community/pull/2335#discussion_r213037163

maybe something like 
// This is alpha feature and may change in the future
// It requires the VolumeSnapshotDataSource alpha feature gate to be enabled and DataSource  must be from VolumeSnapshot

Comment (xing-yang) [reply pkg/apis/core/types.go]: Done

File: pkg/apis/core/types.go
Comment (thockin) [inline pkg/apis/core/types.go]: How do I know if the data init succeeded or not?  Does the provisioner also have to log and event (which disappears) or set an error somewhow?

File: pkg/apis/core/validation/validation.go
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: When you add a second kind or group, you probably want to test them as tuples :)

File: pkg/apis/core/types.go
Comment (xing-yang) [reply pkg/apis/core/types.go]: Added some comments to clarify this.

File: pkg/apis/core/validation/validation.go
Comment (xing-yang) [reply pkg/apis/core/validation/validation.go]: Sure.  supportedDataSourceKinds is defined as sets.NewString(string("VolumeSnapshot”)).  If we added support for a new kind, we can add it to the sets.

**Comment by xing-yang** [comment]
@thockin Addressed your comments.  Please review again.  Thanks!

File: pkg/apis/core/types.go
Comment (jingxu97) [reply pkg/apis/core/types.go]: if provisioner failed to create volume from the data source, the behavior will the same as no data source case (failed to create an empty volume), PV will not be created, and PVC will continue in pending state. The controller will keep retrying to create the volume

**Comment by wackxu** [comment]
you  may need update https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolumeclaim/util.go and update the test like this:
```
	if !utilfeature.DefaultFeatureGate.Enabled(features.VolumeSnapshotDataSource) {
		pvcSpec.VolumeMode = nil
	}
```

**Comment by wackxu** [comment]
verify openapi is failed.
```
/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh
```

**Comment by xing-yang** [comment]
@wackxu That check in https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolumeclaim/util.go seems redundant to me.  If DataSource is not nil and VolumeSnapshotDataSource feature gate is not enabled, user will get an error right away:

        if spec.DataSource != nil && !utilfeature.DefaultFeatureGate.Enabled(features.VolumeSnapshotDataSource) {
                allErrs = append(allErrs, field.Forbidden(fldPath.Child("dataSource"), "VolumeSnapshotDataSource is disabled by feature-gate"))
        }

There is also a similar check for VolumeMode like above, so I don't know why it still needs that code in pkg/api/persistentvolumeclaim/util.go.

**Comment by xing-yang** [comment]
The pull-kubernetes-kubemark-e2e-gce-big test failure was not related to this PR.  I'll re-run it.
Looks like another feature gate change just got merged, so I'll need to resolve the conflict and re-submit.

**Comment by xing-yang** [comment]
@thockin Jing and I discussed about this and want to make the following changes to the DataSource comments.  Let us know what you think.  Thanks!

// This field requires the VolumeSnapshotDataSource alpha feature gate to be enabled and
// currently VolumeSnapshot is the only supported data source.
// If the provisioner does not support the VolumeSnapshot data source, the volume will not be created.
// In the future, we plan to support more data source types and the behavior of the provisioner may change.

**Comment by jingxu97** [comment]
@thockin The comment we have now only describes the provisioner behavior for the only supported data source type, VolumeSnapshot. For this case, the provisioner will create volume and put data in volume at the same time. For the data sources that will be added in the future, it might require an external data populator to handle the data copy. In this case, things will get more complicated. For example, although provisioner does not fully support the data source (it cannot copy data directly), it still creates the volume and let the data populator to handle the data copying. In case of data copy failure, an empty volume might be left there but PVC should be marked as not ready. We need more discussion to finalize the design and cover all the cases. When that feature is ready, we will come back and change the comments here. Please let us know if there is any problem with it. Thank you!

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
